source/mir/bignum/integer.d

```diff
@@ -317,6 +317,190 @@ struct BigInt(size_t maxSize64)
+            BigUIntView!uint quotient = cast(BigUIntView!uint)view.unsigned;
+            data = 0;
+            divm(dividend, divisor, quotient);
+            length = cast(uint)view.normalized.coefficients.length;
+
+            // If this is a negative number, then we should keep the negative sign if we are not negative.
+            // If we are not negative, then we should check if we're dividing by a negative number (,
+            // and apply the negative sign as such.
+            this.sign = rhs.sign ? this.sign != rhs.sign : this.sign;
+        }
+        return this;
+    }
+
+    /++
+    Performs `big %= rhs` operation.
+    Params:
+        rhs = unsigned value to divide by
+    Returns:
+        remainder from division
+    +/
+    ref opOpAssign(string op : "%", size_t size)(UInt!size rhs)
+        @safe pure nothrow @nogc return
+    {
+        import mir.bignum.low_level_view : divm, BigUIntView;
+
+        enum m = (data.length * (size_t.sizeof * 8)), n = (size / (uint.sizeof * 8));
+
+        if (length)
+        {
+            UInt!(m + (size_t.sizeof * 8)) _div;
+            UInt!(data.length * (size_t.sizeof * 8)) q;
+            // shouldn't be unaligned here -- maybe a possibility though
+            _div.data[0 .. $ - 1] = data;
+
+            BigUIntView!uint dividend = cast(BigUIntView!uint)_div.view;
+            BigUIntView!uint divisor = cast(BigUIntView!uint)rhs.view;
+            BigUIntView!uint quotient = cast(BigUIntView!uint)q.view; 
+            divm(dividend, divisor, quotient);
+
+            data = 0;
+            // FIXME: ugly ugly cast
+            (cast(uint[data.sizeof / 4])data)[0 .. dividend.coefficients.length] = dividend.coefficients;
+            length = cast(uint)_div.view.normalized.coefficients.length;
+        }
+        return this;
+    }
+
+    /// ditto
+    ref opOpAssign(string op : "%")(ulong rhs)
+        @safe pure nothrow @nogc return
+    {
+        if (length)
+        {
+            return this.opOpAssign!"%"(UInt!64(rhs));
+        }
+        return this;
+    }
+
+    /// ditto
+    ref opOpAssign(string op : "%")(long rhs)
+        @safe pure nothrow @nogc return
+    {
+        if (length)
+        {
+            ulong div = rhs < 0 ? rhs * -1 : rhs;
+            this.opOpAssign!"%"(div);
+
+            // Add back if it is necessary (XXX: is this even correct??) 
+            if ((this.sign && rhs > 0) || (!this.sign && rhs < 0))
+            {
+                this.opOpAssign!("+")(BigIntView!(const size_t)(UInt!64(div).view, rhs < 0));
+            }
+        }
+        return this;
+    }
+
+    /// ditto
+    ref opOpAssign(string op : "%", size_t rhsMaxSize64)(ref const BigInt!rhsMaxSize64 rhs)
+    {
+        import mir.bignum.low_level_view : divm, BigUIntView;
+
+        enum m = (data.length * (size_t.sizeof * 8)), n = (rhsMaxSize64 / (uint.sizeof * 8));
+
+        if (length)
+        {
+            UInt!(m + (size_t.sizeof * 8)) _div;
```

Comment:
I believe I was getting incorrect results before with length, so I thought it would be easier to switch to just using UInts. I've reverted this and figured out the error (we just need to round to even), let me know what you think.